### ISSUE TITLE
Obtain PECs directly from FAC

### ISSUE THREAD

**Issue by cjperks7** [fsciortino/Aurora#86]
Due to my various needs, I would like there to be more flexibility available in `aurora.radiation.get_local_spectrum()`. I nominate myself to address these needs on appropriate branches to be reviewed by @fsciortino .

NOTE: In my efforts, I will be heavily relying on functions and data files from the Flexible Atomic Code (FAC, https://github.com/flexible-atomic-code/fac). This is because it can satisfy missing components to my needs not covered under OpenADAS. FAC may therefore become a hidden dependency when utilizing specific upgrades I need. Care would be taken for a more general user to have options to avoid this when possible.

### Issue

Aurora currently obtains data files available on OpenADAS or local files in the adf15 format. I am beginning heavy reliance on FAC to calculate photon emissivity coefficients (PECs) under various conditions so would like to integrate the two

### Motivation

- On C-Mod
   - @fsciortino you have already shared with me Ar FAC data files produced by A. Foster in the adf15 file format including all the physics I would want
   - As such, reproducing synthetic spectra for Ar would be a good benchmarking exercise for my edification
   - I would like to explore including information and hence PECs from metals that fall within the spectrum (Mo for C-Mod, W for WEST) to constrain ImpRad modeling. Data for these metals aren't satisfied within OpenADAS
- On SPARC
   - @fsciortino you had calculated for me Kr, Xe adf15 PEC files using ADAS208. But, not all the physics I would want is included (satellites, magnetic transitions) so I need to rerun them. Not that I'm criticizing you here! I wasn't expert enough on He-, Ne-like sequences at the time to properly review. There was enough information for design work, but the team is beginning to transition to more advanced analysis
   - We are also imaging lines from W and other various metals that are not readily available (or data is present but doesn't include enough physics to get the intensity right) in OpenADAS
 - On WEST
     - I am beginning a collaboration with A. Da Ros & D. Vezinet where they are cross-validating electron temperatures inferred via XICS to ECE 
     - Some shot are Lower Hybrid heated so a discrepancy could be due to suprathermal electrons modifying reaction rates
     - As such, non-Maxwellian collisional-radiative modeling is needed to calculate PECs

### Plan of Action

- Understand that when one calculates PECs utilizing FAC, there is two steps
   - The first (and computationally expensive) step is creating data files calculating atomic transition rates/cross-sections
       - This only really needs to be done once per species/charge state and entirely within a FAC workflow
       - This is also the most intense step (hard to automate) per all the physics options so should be handled entirely independent to Aurora
   - The second (and fast) step is then performing collisional-radiative modeling to calculate PECs at a specfic electron density and electron temperature
       - This step is also where you decide the electron energy distribution
       - This step is very short and simple to automate
- As such, I propose that within `aurora.radiation.get_local_spectrum()` a user gets to have a choice of using OpenADAS/adf15 files as normal or perform collisional-radiative modeling with FAC
    - NOTE: This would unavoidably be using FAC within Aurora
    - This would probably amount to a function that completely runs the FAC collisional-radiative model (step 2) given that the user already took the time to run FAC step 1
    - Default would be OpenADAS/adf15 as normal to not perturb other's workflows
- I would probably also integrate this into ImpRad later. Likely more important to transport inferences I want to do on WEST where there are a lot of polluting W lines imaged on the He-, H-like Ar systems
- NOTE: I have had a discussion for C. Johnson to instead make an upgrade in ColRadPy to read FAC step 1 data files to then use the ColRadPy collisional-radiative modeling functions I know you've already implemented into Aurora. I feel like just sticking with FAC step 2 would be more direct. The FAC step 2 function would probably do about the same things as your ColRadPy function

@fsciortino any opinions?

**Comment by fsciortino** [comment]
This would also be a great contribution, @cjperks7 ! One request: please keep the code flexible. Most users don't have a clue of what FAC is, and they shouldn't have to find out. If you decide to make a whole module that helps with FAC operation, similar to the ADAS module or the AMJUEL module, that would be fantastic! Several tools (like those that you mentioned) should be generalized to be able to use rates from whatever database or tool, in principle.

Fundamentally, ADF15 is just a format. It doesn't matter what produced the file in that format. So, what I've done a few times was to simply use ADF15 as a communication standard.

**Comment by cjperks7** [comment]
Yeah, I think the problem, as was in ColRadPy, is reading files from a different database. Both shouldn't have problems to use the rates in collisional-radiative modeling. The one benefit of FAC over ColRadPy though is I think the former can due non-Maxwellian electrons per conversations I've had with Curt. Though ColRadPy has its own specific strengths. 

My big hangup with not just writing FAC output into ADF15 format myself to make this whole point moot is that I don't have on hand something to due that conversation and don't think I want to invest my time writing one. You wouldn't happen to have such a conversation? It would be good if any arbitrary database could be made into a standard format (ADF15)

Either way, I pretty sure all that's done to make an ADF15 is do a bunch of CR model runs looping over electron density/temperature so I figure why not just do it per whatever profiles you instead. But of course that may be slow.  

Feel free to point collaborators my way if they have interest in FAC.

**Comment by fsciortino** [comment]
I think that you mean "conversion", not "conversation" :)

Adam Foster at Harvard has such a function to write data in ADF15 format. If you shoot him an email, I think that he'll be happy to help!

**Comment by cjperks7** [comment]
@fsciortino Thanks for the tip! Adam was very quick to respond and be helpful. None of this needs to be in Aurora and I can handle formatting FAC files externally.
